internal/config/instance.go

```diff
@@ -29,21 +33,112 @@ type Instance struct {
 	Exit          func(code int)
 }
 
-// New creates a new config instance
-func New(localPath string) *Instance {
+func new(localPath string) *Instance {
 	instance := &Instance{
+		viper:     viper.New(),
 		localPath: localPath,
 		Exit:      os.Exit,
 	}
 
-	instance.ensureConfigExists()
-	instance.ensureCacheExists()
-	instance.ReadInConfig()
-	instance.readInstallSource()
+	instance.Reload()
 
 	return instance
 }
 
+// Reload reloads the configuration data from the config file
+func (i *Instance) Reload() {
+	i.ensureConfigExists()
+	i.ensureCacheExists()
+	i.ReadInConfig()
+	i.readInstallSource()
+}
+
+func configPathInTest() (string, error) {
+	localPath, err := ioutil.TempDir("", "cli-config")
+	if err != nil {
+		return "", fmt.Errorf("Could not create temp dir: %w", err)
+	}
+	err = os.RemoveAll(localPath)
+	if err != nil {
+		return "", fmt.Errorf("Could not remove generated config dir for tests: %w", err)
+	}
+	return localPath, nil
+}
+
+// New creates a new config instance
+func New() *Instance {
+	localPath := os.Getenv(C.ConfigEnvVarName)
+
+	if condition.InTest() {
+		var err error
+		localPath, err = configPathInTest()
+		if err != nil {
+			// panic as this only happening in tests
+			panic(err)
+		}
+	}
+
+	return new(localPath)
+}
+
+// NewWithDir creates a new instance at the given directory
+func NewWithDir(dir string) *Instance {
+	return new(dir)
+}
+
+// Get returns the default configuration instance
+func Get() *Instance {
+	if defaultConfig == nil {
+		defaultConfig = New()
+	}
+	return defaultConfig
+}
+
+// WriteConfig writes the configuration
+func (i *Instance) WriteConfig() error {
```

Comment:
I think we can drop this, from a quick search the code that uses `viper.WriteConfig` should probably be using `config.Save`.